bundles/org.openhab.binding.opengarage/src/main/java/org/openhab/binding/opengarage/internal/VariableDelayPoller.java

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (c) 2010-2022 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.opengarage.internal;
+
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Variable poller will schedule a task to run periodically, but will allow the
+ * schedule to be modified on-the-fly
+ *
+ * @author Tim Harper - Initial contribution
+ */
+public class VariableDelayPoller {
```

Comment:
Please add `@NonNullByDefault` and then add `@Nullable` when required in the rest of the class.

I have the strange feeling you made stuff complex with that class while not necessary, but let's assume you have a good reason I do not understand.

Reply:
Yes, I have a hunch you are right, and felt this code needed justification. I wrote up a long response to justify it, but after writing it all, I think the reason was excessive paranoia about concurrency and potential deadlocks. I think a simple approach will be fine.

justification I wrote:

 It may be a case of being excessively paranoia about multi-threading and not entirely sure what I can trust to not create a deadlock scenario. I can probably be talked out of this.

With the previous implementation, we would simply run the poll on a loop and be done with it. Ezpz. Scheduler handles it.

- 00:01 - Poll
- 00:30 - Door open command
- 00:31 - Poll
- 01:01 - Poll
- 01:31 - Poll


With the door transition state change, we need to additionally update the state in response to a door state command (in order to convey the transition state wanted by HomeKit), and then, again, when the door transition is completed. Let's assume a transition time of 17 seconds. Here's what the schedule would look like without my variable poller:

- 00:01 - Poll
- 00:30 - Door open command
- 00:30 - "Poll" (in response to door command)
- 00:31 - Poll (on schedule)
- 00:47 - Poll (in response to earlier door command, to publish door transition completion)
- 01:01 - Poll
- 01:31 - Poll

Now, if two of these poll methods happen to run concurrently, then it could cause some trouble. Say 00:30 poll invocation starts, and while it's running, the 00:31 poll invocation runs and completes before the 00:30 poll invocation completes. We'd end up with the state of 00:30 overriding the state of 00:31. Probably relatively harmless? I don't know. Then, on shutdown, we have extra future we have to potentially cancel.

If we put synchronized around the pollStatus loop, we risk clogging the thread pool with up to 3 threads because one thread is taking 30 seconds to time out.

I don't know, the more I'm trying to explain this the less justified I feel it really is. I'm going to try and delete it.

Reply:
I removed it, I'm pretty confident it'll be fine just using synchronized and an extra one-of scheduled task.